### ISSUE TITLE
Some stuff don't work on Cervantes 2013

### ISSUE THREAD

**Issue by ZapperDJ** [koreader/koreader#5084]
KOReader version: 2019.06
* Device: Cervantes 2013

#### Issues
1. Networking is unavailable, accessing the menu crashes KOReader
2. Whenever go to to Exit > Suspend on the KOReader menu, a message pops up with the text Sleeping and then KOReader exits

#### Logs
[crash_wifi.log](https://github.com/koreader/koreader/files/3294354/crash_wifi.log)
[crash-suspend.log](https://github.com/koreader/koreader/files/3294365/crash-suspend.log)

**Comment by pazos** [comment]
Hi @ZapperDJ, thanks for reporting!

I see two local functions failing. So I need that you check these two commands from Telnet (without KOReader running AND with network enabled)

The first -> `/usr/bin/ntxinfo /dev/mmcblk0 | grep pcb | cut -d ':' -f2", "r"`
The second -> `ls /sys/class/net/eth0/carrier || ls /sys/class/net` 

Probably both issues are easy to fix, but I had no report about C2013 until now!

**Comment by ZapperDJ** [comment]
First command output is 33
Secondo mommand outputs /sys/class/net/eth0/carrier. This makes no sense to me, maybe you meant cat /sys/class/net/eth0/carrier? That outputs 1 by the way

**Comment by pazos** [comment]
no, I wanted to make you have the same network interface, eth0

**Comment by pazos** [comment]
@ZapperDJ: are you still interested?

**Comment by ZapperDJ** [comment]
Yes, sorry I missed your reply on this bug report. I tried both solutions and I'm happy to say that option 2 fixed both the networking and suspending issues. However, there is a minor bug still present. After connecting to a wifi network, the checkbox for the option network -> wifi connection remains unchecked, and the network info option is disabled

**Comment by pazos** [comment]
thanks for reporting!! Will look at it (and the other cervantes stuff) in a few weeks. Now I'm fully into the android side of the things!

**Comment by frio** [comment]
The wifi thing is something that happens on the 3 as well; I have a fix that I've applied manually and forgotten to push upstream.  I'll dig that up tonight :).

**Comment by pazos** [comment]
> The wifi thing is something that happens on the 3 as well; I have a fix that I've applied manually and forgotten to push upstream. I'll dig that up tonight :).

@frio: please go ahead :+1: I'm actually using my C4 for reading and I would prefer leaving it as-is.

**Comment by ZapperDJ** [comment]
> The wifi thing is something that happens on the 3 as well; I have a fix that I've applied manually and forgotten to push upstream. I'll dig that up tonight :).

Could you please share that fix? Is the last issue I have pending on my Cervantes

**Comment by pazos** [comment]
@ZapperDJ: while we wait you can try to solve it yourself. It should be pretty easy now that you're comfortable with dev tools:

1. connect your device with telnet (KOReader not running yet).
2. Create a new lua script and copy there the current `isConnected` function and a print statement. It should look like:

```bash
#!/mnt/private/koreader/luajit

local function isConnected()
    local file = io.open("/sys/class/net/eth0/carrier", "rb")
    if not file then return 0 end
    local carrier = tonumber(file:read("*all")) or 0
    file:close()
    return carrier
end

print(isConnected())
```

3. set your script executable (asumming is it stored in /mnt/private with the name test.lua): `chmod +x /mnt/private/test.lua`
4. now run your script: `/mnt/private/test.lua`

It should return 0 unless you're associated with an access point. In that case it will return 1.

5. Modify the script to return the value of /sys/class/net/eth0/carrier or 0 if that sysfile doesn't exist.

**Comment by frio** [comment]
Sorry; I'll try and get onto it tonight :).  I was naughty and modified it on-device, and resurrecting the maze of telnet and related skullduggery to retrieve what I did is a little more arduous than it could be.

**Comment by pazos** [comment]
@ZapperDJ @frio Please try with this function:

```lua
-- returns 1 if we're associated with the AP, 0 otherwise
local function isConnected()
    -- read carrier state from sysfs (for eth0)
    local file = io.open("/sys/class/net/eth0/carrier", "rb")

    -- file exists while wifi module is loaded.
    if not file then return 0 end

    -- 0 means not connected, 1 connected
    local out = file:read("*all")
    file:close()

    -- strip NaN from file read (ie: line endings, error messages)
    if type(out) ~= "number" then
        carrier = tonumber(out)
    else
        carrier = out
    end

    -- finally return if we're connected or not
    if type(carrier) == "number" then
        return carrier
    else
        return 0
    end
end

```

**Comment by ZapperDJ** [comment]
OK tried replacing the function with the code you provided and now it works. There's only two odd things I noticed:
1 Isn't there a way to make KOReader forget wifi network password? 
2 You must disconnect wifi on KOReader first if you want to switch to BQ reader app, otherwise wifi won't work in BQ app

**Comment by pazos** [comment]
1. Nope from the UI. You can edit a file under settings but not sure right now how it is named (probably something network or something wpa). Won't fix
2. Thanks. The other day I was under the impression that on Cervantes we release wifi but I was wrong. Easy fixable

**Comment by frio** [comment]
Ah, finally, here we go (... recompiling a kernel with `usbnet` later...): I modified `Cervantes:initNetworkManager`:

```
-- wireless
function Cervantes:initNetworkManager(NetworkMgr)
    function NetworkMgr:turnOffWifi(complete_callback)
        logger.info("Cervantes: disabling WiFi")
        self:releaseIP()
        os.execute("./disable-wifi.sh")
        if complete_callback then
            complete_callback()
        end
    end
    function NetworkMgr:turnOnWifi(complete_callback)
        logger.info("Cervantes: enabling WiFi")
        os.execute("./enable-wifi.sh")
        self:reconnectOrShowNetworkMenu(complete_callback)
    end
    NetworkMgr:setWirelessBackend("wpa_supplicant", {ctrl_interface = "/var/run/wpa_supplicant/eth0"})
    function NetworkMgr:obtainIP()
        os.execute("./obtain-ip.sh")
    end
    function NetworkMgr:releaseIP()
        os.execute("./release-ip.sh")
    end
    function NetworkMgr:restoreWifiAsync()
        os.execute("./restore-wifi-async.sh")
    end
    function NetworkMgr:isWifiOn()
        local completed, status = pcall(isConnected)
        return completed and status == 1
    end
end
```

Note the `pcall` at the end.  I haven't tested your fix yet (thanks!), but it's almost definitely better than my cheap hack ;).
